### ISSUE TITLE
_all segment in mappings -> How to create?

### ISSUE THREAD

**Issue by jjradke** [elastic/elasticsearch-net#1511]
Good morning, 

I am using this article as the basis for setting up my index:

http://blog.qbox.io/multi-field-partial-word-autocomplete-in-elasticsearch-using-ngrams

I am able to get almost everything set up properly using Nest.  The one piece I cannot figure out is the _all property in the mappings:

   "mappings": {
      "movies": {
         "_all": {
            "index_analyzer": "nGram_analyzer",
            "search_analyzer": "whitespace_analyzer"
         },
         .....

I tried doing this:

                var ix = new IndexSettings();
                List<RootObjectMapping> mappings = new List<RootObjectMapping>();

                RootObjectMapping rom = new RootObjectMapping();
                rom.Name = "_all";
                rom.Enabled = true;
                rom.IncludeInAll = true;

                AllFieldMapping afm = new AllFieldMapping();
                afm.Enabled = true;
                afm.IndexAnalyzer = "nGram_analyzer";
                afm.SearchAnalyzer = "whitespace_analyzer";
                rom.AllFieldMapping = afm;
                mappings.Add(rom);

                ix.Mappings = mappings;

                IndexNameMarker imark = new IndexNameMarker();
                imark.Name = indexName;
                CreateIndexRequest xir = new CreateIndexRequest(imark);
                xir.IndexSettings = ix;

                client.CreateIndex(xir);

And use those index settings to create the index (which will also use the settings in the annotations on the class, I hope, to set the mapings).

I see the annotations from the class work properly, but I never see _all item like in the example from the article.

Can someone help me understand how to get this segment created properly via Nest?

Thanks,

Jim

**Comment by gmarz** [comment]
Hey @jjradke 

`RootObjectMapping` is the root object for the type itself and defines a mapping for a particular type, which in this case (if following the linked article) would be `movies`.

```csharp
 RootObjectMapping rom = new RootObjectMapping();
 rom.Name = "_all";
```

`Name` here would be the name of the type `movies`, not `_all`.  To configure and enable the `_all` field, you just need to set the `AllFieldMapping` property as you're already doing.

Here's a shortened example of how you would setup the mapping described in that link.  Perhaps this might clear things up:

```csharp
var moviesMapping = new RootObjectMapping
{
	Name = "movies",
	AllFieldMapping = new AllFieldMapping
	{
		Enabled = true,
		IndexAnalyzer = "nGram_analyzer",
		SearchAnalyzer = "whitespace_analyzer"
	},
	Properties = new Dictionary<PropertyNameMarker, IElasticType>
	{
		{ "addToCartUrl", new StringMapping { Index = FieldIndexOption.No, IncludeInAll = false } },
		{ "format", new StringMapping { Index = FieldIndexOption.NotAnalyzed } },
        // snip...
	}
};

var indexSettings = new IndexSettings
{
	Settings = // snip...
	Mappings = new List<RootObjectMapping> { moviesMapping }
};

var createIndexRequest = new CreateIndexRequest("blurays")
{
	IndexSettings = indexSettings
};
```

BTW, we're currently in the process of [rewriting the mapping API for NEST 2.0](https://github.com/elastic/elasticsearch-net/tree/fix/2.0-mappings), so this will be much more intuitive.


Hope that helps!

**Comment by jjradke** [comment]
Will it merge into the properties everything that it finds in the ElasticType annotations in the class for that object, or do I have to hard-code all the properties in the code?

That did get the _all segment working, thanks!

**Comment by jjradke** [comment]
I meand ElasticProperty annotations in the last comment!!  Sorry about that...

**Comment by gmarz** [comment]
@jjradke in order to leverage the `ElasticType` attribute, you need to use the preferred fluent API in order to call `MapFromAttributes()`.  Perhaps this is something we can also fix/unify as part of the 2.0 mapping rewrite.

Above example using the fluent API, assuming you have a POCO named `Movie` with the `ElasticType` attribute defined on its properties:

```csharp
client.CreateIndex("blurays", c => c
	.AddMapping<Movie>(m => m
		.MapFromAttributes()
		.AllField(a => a
			.Enabled()
			.IndexAnalyzer("nGram_analyzer")
			.SearchAnalyzer("whitespace_analyzer")
		)
	)
);
```

**Comment by jjradke** [comment]
Greg,

That looks even easier!  I will try it..

Thanks for all of your help!

I really do like NEST, it has worked well for me so far on my project; and
I am a beginner in ElasticSearch.

The combination of both tools seem very powerful.

Is 2.0 far along, or will it be a while before we see the 2.0 release?

Thanks,

Jim

**Comment by gmarz** [comment]
Awesome, glad to hear you like the client!

Yea, the fluent API is the more preferred way of using NEST and definitely tends to get more love in `1.x`.  The object initializer syntax was added after the fact, so there are still some gaps in functionality- all which we plan to have addressed in `2.0`.

As for timeline, can't really commit to anything, but it's definitely our priority at the moment.  Stay tuned :).
